### DISCUSSION TITLE
Opencv, cant open img.

### DISCUSSION THREAD

**Discussion by snirbenyosef** [coder/code-server#2090] [Q&A]
HI, i׳m trying to use cv2.imgshow on a vs code online my code: 
path = r'6.png' 
image = cv2.imread(path) 
window_name = 'image' 
cv2.imshow(window_name, image) 
cv2.waitKey(0)
cv2.destroyAllWindows()

but all i get is : cannot connect to X server

how can i use cv2.imgshow in vs code online ? and ideas ?

**Comment by code-asher** [comment]
Looks like it requires a running X server. You'd need to install and run it on the remote machine then connect to it in some way (for example VNC).

**Reply by snirbenyosef** [reply]
could it be done with docker?
my main work is with opencv, and the code server is working on amazon linux vm.

**Reply by code-asher** [reply]
I've done it with Docker before. I use `x11vnc` and `noVNC` because I don't want to install a VNC client (noVNC lets you use a browser). In my case I also had to use `xvfb-run`.

Here are some Docker images that show how to do this:
https://github.com/ConSol/docker-headless-vnc-container
https://github.com/fcwu/docker-ubuntu-vnc-desktop

Some other links that might be helpful:
https://stackoverflow.com/questions/16296753/can-you-run-gui-applications-in-a-docker-container
https://qxf2.com/blog/view-docker-container-display-using-vnc-viewer/

**Reply by snirbenyosef** [reply]
thank you!
just making sure.

i use a docker to run my vscode online. and if i will create a new docker with your suggestions i should have opencv working properly?

**Reply by code-asher** [reply]
Yup, if you create a Docker image with all the X stuff opencv should work.

You might want to create a single combined image for both code-server and opencv, otherwise if you try to run opencv from code-server it won't work (for example from code-server's terminal).

But if you edit code in code-server and you run the code in a separate terminal then you could have them as separate Docker containers with the same directory mounted.

**Reply by snirbenyosef** [reply]
what do you mean by both code-server and opencv combined?
opencv is a python library no ?

in my vs code i can import cv2 etc

**Reply by snirbenyosef** [reply]
how i should configure the x server in the vscode terminal, to link it to the docker with x stuff ?

**Reply by code-asher** [reply]
Oops, sorry about that. I mean an image with both code-server and X.

In the Docker file you could set the `DISPLAY` environment variable before starting code-server. Then code-server and any other processes running will be able to connect to the X display.

Or you can set `DISPLAY` in your shell configuration (for example `.bashrc`). That way when you spawn a new terminal anything inside that terminal will be able to connect to the X display.

**Reply by snirbenyosef** [reply]
i'm sorry but i was unable to understand how to accomplished it.
can you please tell me in steps how can i use docker image in order to use opencv and cv2 with vs code online?
my code-server(vscode online) runs on a docker also and opencv installed on that docker.
i just need a docker to show the cv2 img. like i am using normal pc and not online.
this is will be very helpful to my company and my self .
thank you!

**Reply by code-asher** [reply]
It might be difficult to find the time but if I can I'll try to put something together.

**Reply by itaynvn** [reply]
@code-asher i was looking for this as well, would appreciate it very much :)

**Comment by sr229** [comment]
OpenCV requires a working windowing system since it needs to output of the camera to a window. Please install X Server before dealing with APIs like OpenCV.

**Reply by snirbenyosef** [reply]
could it be done with docker?
my main work is with opencv, and the code server is working on amazon linux vm.

**Reply by sr229** [reply]
Please refer to @code-asher's answer.
